File: AcWing/基础算法/quicksort.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-
# @Time  : 17/1/2022 下午 6:50
# @Author: xiaoni
# @File  : quicksort.py

'''
    任然存在
'''

def quicksort(aa, low, high):
    if low >= high:
        return
    p = aa[low]
    lowtem = low
    hightem = high

    while lowtem < hightem:
        while lowtem < hightem and aa[hightem] >= p:
            hightem = hightem - 1
        aa[lowtem] = aa[hightem]
        while lowtem < hightem and aa[lowtem] <= p:
            lowtem = lowtem + 1
        aa[hightem] = aa[lowtem]

    aa[lowtem] = p
    quicksort(aa, low, lowtem - 1)
    quicksort(aa, lowtem + 1, high)


if __name__ == "__main__":
    n = int(input())
    a = [int(x) for x in input().split()]
    quicksort(a, 0, len(a) - 1)
    for i in range(len(a) ):
        print(a[i], end='')
        print(" ", end='')

File: AcWing/basic algorithm/quicksort.cpp

#include <iostream>

const int N = 1e6 + 10;
int n;
int q[N];
void quicksort(int q[], int low, int high)
{
    if (low >= high)
        return;

    int temp = q[low];
    int i = low;
    int j = high;
    while (i < j)
    {
        while (i < j && q[j] >= temp)
            j--;
        q[i] = q[j];
        while (i < j && q[i] <= temp)
            i++;
        q[j] = q[i];
    }
    q[i] = temp;
    quicksort(q, low, i - 1);
    quicksort(q, i + 1, high);
}
int main()
{
    scanf("%d", &n);
    for (int i = 0; i < n; i++)
    {
        scanf("%d", &q[i]);
    }
    quicksort(q, 0, n - 1);
    for (int i = 0; i < n; i++)
    {
        printf("%d", q[i]);
    }
}
